### ISSUE TITLE
TFVC scorch deletes too much when no root workspace mapping

### ISSUE THREAD

**Issue by JKennedy24** [Microsoft/vsts-agent#438]
## **Agent version and platform**
v2.103.1, Windows 10
<br>
## **VSTS type and version**
VisualStudio.com
## **What's not working?** 
Setting the Clean = True parameter on my build removes all my files on second run but does not redownload
## **Agent and Worker's diag log**
[2_Get sources.txt](https://github.com/Microsoft/vsts-agent/files/381000/2_Get.sources.txt)

**Comment by bryanmacfarlane** [comment]
It looks like it did.

...
2016-07-25T07:48:24.4443179Z Deleting wp8
2016-07-25T07:48:24.4443179Z Deleting wp81
2016-07-25T07:48:24.4911919Z ##[command]tf vc get /version:8053 /recursive /overwrite C:\agent\_work\1\s /loginType:OAuth /login:.,******** /noprompt
2016-07-25T07:48:25.7400456Z C:\agent\_work\1\s\F\FS\FieldStrikeMove.Business:
2016-07-25T07:48:25.7400456Z Replacing FieldStrikeMove.Business.csproj
2016-07-25T07:48:25.7556667Z Replacing packages.config
...

**Comment by ericsciple** [comment]
@JKennedy24 Did it delete *all* of your files? It ran scorch which should just delete files not under source control. At that point, when `tf get` runs it is just pulling incrementally.

Are you saying `tf scorch` deleted more than it should have? Or `tf get` missed something?

**Comment by JKennedy24** [comment]
After running the `Get Sources` step of my build definition nothing was left in my agent\work\1 folder. Just loads of empty folders.

All the files it deleted are files which should be there and I require to build my project. 

Take for example the file BaseAudit.cs which is a file that is currently in source control and part of my project (not deleted).

we see in the log the line:

2016-07-25T07:48:22.8036890Z Deleting BaseAudit.cs

but then it never re downloads that file. This only happens when clean = true on the build def and running build agent v2.103.1

Let me know if you need any more info or more detailed logs

**Comment by JKennedy24** [comment]
its almost as if `tf vc get` is only `Replacing` files that have changed since the last build by the build agent and all other files are being deleted

**Comment by JKennedy24** [comment]
here are my `Get Sources` logs from the build agent before this version
[1_Get sources.txt](https://github.com/Microsoft/vsts-agent/files/385938/1_Get.sources.txt)

You can see its not deleting Source Control items only items that arent in source control.

From that I think it definitely shows scorch is deleting too much when clean = true

**Comment by ericsciple** [comment]
Thanks @JKennedy24 I've got a repro. Sorry I didn't notice this when I made the previous "fix" for scorch to work with unrooted workspaces :(

Setting `build.clean` to `source` is a workaround (drops workspace and re-clones). Fyi - in an attempt to find a better workaround, I tried adding a single-level root mapping ($/*) but the workspace is getting dropped every time (investigating that now).

**Comment by ericsciple** [comment]
2.104.1 release will contain these related changes: https://github.com/Microsoft/vsts-agent/pull/453

- I reverted the behavior so that scorch will fail when the root sources directory (e.g. `c:\a\1\s`) is not mapped. The difference now is that the exception will be caught and logged as a warning. The agent will proceed to drop the workspace and create a new workspace.
- A better workaround (which will enable scorch to succeed properly) is to map the root sources to something, e.g. even a single-level map such as `$/*`. I fixed a bug where `$/*` wasn't matched properly when attempting to find a workspace to reuse.

I'm tracking the scorch issue here: https://github.com/Microsoft/vsts-agent/issues/398
